Sources/NIOWebSocket/NIOWebSocketFrameAggregator.swift

```diff
@@ -0,0 +1,138 @@
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let frame = unwrapInboundIn(data)
+        do {
+            switch frame.opcode {
+            case .continuation:
+                guard let firstFrameOpcode = bufferedFrames.first?.opcode else {
+                    throw Error.didReceiveFragmentBeforeReceivingTextOrBinaryFrame
+                }
+                try bufferFrame(frame)
+                
+                guard frame.fin else { break }
+                // final frame received
+                
+                let aggregatedFrame = aggregateFrames(frames: bufferedFrames, opcode: firstFrameOpcode)
+                clearBuffer()
+                
+                context.fireChannelRead(wrapInboundOut(aggregatedFrame))
+            case .binary, .text:
+                if frame.fin {
+                    guard bufferedFrames.isEmpty else {
+                        throw Error.receivedNewFrameWithoutFinishingPrevious
+                    }
+                    // fast path: no need to check any constraints nor unmask and copy data
+                    context.fireChannelRead(data)
+                } else {
+                    try bufferFrame(frame)
+                }
+            default:
+                // control frames can't be fragmented
+                context.fireChannelRead(data)
+            }
+        } catch {
+            // free memory early
+            clearBuffer()
+            context.fireErrorCaught(error)
+        }
+    }
+    
+    private func bufferFrame(_ frame: WebSocketFrame) throws {
+        guard bufferedFrames.isEmpty || frame.opcode == .continuation else {
+            throw Error.receivedNewFrameWithoutFinishingPrevious
+        }
+        guard frame.fin || frame.length >= minNonFinalFragmentSize else {
+            throw Error.nonFinalFragmentSizeIsTooSmall
+        }
+        guard bufferedFrames.count < maxAccumulatedFrameCount else {
+            throw Error.tooManyFragments
+        }
+        
+        // if this is not a final frame, we will at least receive one more frame
+        guard frame.fin || (bufferedFrames.count + 1) < maxAccumulatedFrameCount else {
+            throw Error.tooManyFragments
+        }
+        
+        bufferedFrames.append(frame)
+        
+        guard accumulatedFrameSize <= maxAccumulatedFrameSize else {
+            throw Error.accumulatedFrameSizeIsTooLarge
+        }
+    }
+    
+    private func aggregateFrames(frames: [WebSocketFrame], opcode: WebSocketOpcode) -> WebSocketFrame {
+        var dataBuffer = ByteBuffer()
+        dataBuffer.reserveCapacity(minimumWritableBytes: accumulatedFrameSize)
+        
+        for frame in bufferedFrames {
+            var unmaskedData = frame.unmaskedData
+            dataBuffer.writeBuffer(&unmaskedData)
+        }
+        
+        return WebSocketFrame(fin: true, opcode: opcode, data: dataBuffer)
+    }
+    
+    private func clearBuffer() {
+        bufferedFrames = []
```

Comment:
To reduce allocator traffic it might be best to preserve the existing allocated capacity where we can: maybe use `bufferedFrames.removeAll(keepingCapacity: true)`?

Reply:
Done. 
`accumulatedFrameSize` is currently  also not cached and it just iterates over all buffered frames to get the current accumulated size. We should probably also cache the sum to improve performance. WDYT?

Reply:
Yup, I think that's a good idea.